Relative test result file path not being applied correctly

**Description**

Hi, first of all, great work on the extension! It really helps a lot when running on Linux where the VSTest task is not supported.

When setting the properties `Test Results Directory` and `Test Result Files`, I've encountered an issue in the way the path is combined.

![image](https://github.com/user-attachments/assets/c3549732-12ea-46b5-9d6c-ed85a5a7c425)

The `TargetTestResultsFolderSearch` variables resolves to `$(Agent.TempDirectory)/TestResults/Search` and that resolves to `/home/svc_prc_inttest/azagent/_work/_temp/TestResults/Search/mstest`

The result is
```
Not found testResultFile(s): /home/svc_prc_inttest/azagent/_work/_temp/TestResults/Search/mstest/home/svc_prc_inttest/azagent/_work/r1/a/SearchTests.trx
```

Here you can see the `Test Results Directory` property is combined with `/home/svc_prc_inttest/azagent/_work/r1/a` (which is the default working directory I assume) and then with the `Test Result Files` property (the file name).

**Environment Details**

I am using Azure DevOps Server 2022.2 RC (AzureDevopsServer_20240423.2) with a deployment job running on a linux agent.

- build-agent: self-hosted (linux)
- test-format: mstest
- test-lanugage: C#
- pipeline-format: Designer-based (Classic)

**Pipeline Output**

```
2024-11-07T06:33:07.6158532Z ##[debug]testResultFormat=mstest
2024-11-07T06:33:07.6163521Z ##[debug]testResultDirectory=/home/svc_prc_inttest/azagent/_work/_temp/TestResults/Search/mstest
2024-11-07T06:33:07.6168392Z ##[debug]testResultFiles=/home/svc_prc_inttest/azagent/_work/r1/a/SearchTests.trx
2024-11-07T06:33:07.6171479Z ##[debug]check path : /home/svc_prc_inttest/azagent/_work/_temp/TestResults/Search/mstest/home/svc_prc_inttest/azagent/_work/r1/a/SearchTests.trx
2024-11-07T06:33:07.6192983Z ##[debug]task result: Failed
2024-11-07T06:33:07.6264184Z ##[error]Not found testResultFile(s): /home/svc_prc_inttest/azagent/_work/_temp/TestResults/Search/mstest/home/svc_prc_inttest/azagent/_work/r1/a/SearchTests.trx
```

**To Reproduce**
Steps to reproduce the behaviour:
1. Define a classic release pipeline
2. Add the `Publish Test Plan Results` task
3. Set the required properties
4. Set `Test Results Directory` to an absolute path
5. Set `Test Result Files` to a trx file name

**Expected behaviour**

The path is combined correctly and resolves to `<Test Results Directory>/<Test Result Files>`, in my case `/home/svc_prc_inttest/azagent/_work/_temp/TestResults/Search/mstest/SearchTests.trx`

**Workaround**

As a workaround I currently use this:

![image](https://github.com/user-attachments/assets/b9a8b315-6dc0-43ad-b1cc-e573b4368f9e)

This resolves correctly:

```
2024-11-07T06:53:28.6754761Z ##[debug]testResultDirectory=undefined
2024-11-07T06:53:28.6756562Z ##[debug]testResultFiles=/home/svc_prc_inttest/azagent/_work/_temp/TestResults/Search/mstest/SearchTests.trx
2024-11-07T06:53:28.6757867Z ##[debug]check path : /home/svc_prc_inttest/azagent/_work/_temp/TestResults/Search/mstest/SearchTests.trx
```

--- a/PublishTestPlanResultsV1/TaskParameters.ts
+++ b/PublishTestPlanResultsV1/TaskParameters.ts
@@ -35,21 +35,7 @@ export function getFrameworkParameters(): TestFrameworkParameters {
   tl.debug("reading TestFrameworkParameters from task inputs.");
 
   let testResultFormat = tl.getInput("testResultFormat", true);
-  let testResultFolder = tl.getInput("testResultDirectory", false);
-  let testResultFiles = tl.getDelimitedInput("testResultFiles", ",", true)
-    .map(file => {
-      if (testResultFolder) {
-        return path.join(testResultFolder, file);
-      }
-      return file;
-    })
-    .filter(file => {
-      // verify that file exists
-      if (file.indexOf('**') == -1) {
-        tl.checkPath(file, "testResultFile(s)");
-      }
-      return true;
-    });
+  let testResultFiles = getTestFiles();
 
   return new TestFrameworkParameters(testResultFiles, testResultFormat!.toLowerCase()); 
 }
@@ -76,4 +62,33 @@ export function getPublisherParameters() : TestRunPublisherParameters {
   const dryRun = tl.getBoolInput("dryRun", false);
   const testRunTitle = tl.getInput("testRunTitle", false) ?? "PublishTestPlanResult";
   return new TestRunPublisherParameters(collectionUri as string, accessToken as string, dryRun, testRunTitle, buildId as string);
+}
+
+function getTestFiles() : string[] {
+  
+  let testResultFolder = tl.getInput("testResultDirectory", false);
+  if (testResultFolder == undefined) {
+    // System.DefaultWorkingDirectory:
+    // - build pipelines: "C:\agent\work\1\s" equivalent to "$(Build.SourcesDirectory)"
+    // - release pipelines: "C:\agent\work\r1\a" equivalent to "$(System.ArtifactsDirectory)"
+    testResultFolder = tl.getVariable("SYSTEM_DEFAULTWORKINGDIRECTORY") as string;
+  }
+
+  let testResultFiles = tl.getDelimitedInput("testResultFiles", ",", true)
+    .map(file => {
+      // merge relative paths with the testresult folder
+      if (!path.isAbsolute(file)) {
+        return path.join(testResultFolder, file);
+      }
+      return file;
+    })
+    .filter(file => {
+      // verify that file exists
+      if (file.indexOf('**') == -1) {
+        tl.checkPath(file, "testResultFile(s)");
+      }
+      return true;
+    });
+
+  return testResultFiles;
 }

--- a/PublishTestPlanResultsV1/Test-ExtensionLocally.ps1
+++ b/PublishTestPlanResultsV1/Test-ExtensionLocally.ps1
@@ -90,6 +90,10 @@ if ($BuildId) {
   [System.Environment]::SetEnvironmentVariable("BUILD_BUILDID", $BuildId)
 }
 
+if (!$env:SYSTEM_DEFAULTWORKINGDIRECTORY) {
+  [System.Environment]::SetEnvironmentVariable("SYSTEM_DEFAULTWORKINGDIRECTORY", (Resolve-Path .))
+}
+
 
 if ($DebugMode.IsPresent) {
   [System.Environment]::SetEnvironmentVariable("SYSTEM_DEBUG", "true");

--- a/PublishTestPlanResultsV1/task.json
+++ b/PublishTestPlanResultsV1/task.json
@@ -112,7 +112,7 @@
         },
         {
             "name": "testResultFiles",
-            "type": "filePath",
+            "type": "string",
             "required": true,
             "label": "Test Result Files",
             "groupName": "general",

--- a/PublishTestPlanResultsV1/test/TaskParameters.specs.ts
+++ b/PublishTestPlanResultsV1/test/TaskParameters.specs.ts
@@ -96,6 +96,10 @@ describe('TaskParameters', () => {
       invalidFiles.push( path.join("xunit", "invalidFile1.xml") );
     });
 
+    beforeEach(() => {
+      util.setSystemVariable("System.DefaultWorkingDirectory", __dirname);
+    })
+
     it('Should require testResultFormat to be provided', () => {
       // arrange
       // leave testResultFormat empty
@@ -121,7 +125,7 @@ describe('TaskParameters', () => {
       util.shouldThrow( () => TaskParameters.getFrameworkParameters(), "Input required: testResultFiles");
     });
 
-    it('Should verify that result files are valid', () => {
+    it('Should verify that absolute result files are valid', () => {
       // arrange
       util.setInput("testResultFormat", "xUnit");     
       util.setInput("testResultFiles", validFiles[0]);
@@ -151,16 +155,30 @@ describe('TaskParameters', () => {
       expect(parameters.testFiles[0]).to.eq(validFiles[0]);
     })
 
+    it('Should use default working directory as base results folder when relative path is provided', () => {
+      // arrange
+      util.setInput("testResultFormat", "xUnit");
+      util.setInput("testResultFiles", "data/xunit/xunit-1.xml");
+      util.loadData();
+
+      // act
+      require(tp);
+      var parameters = TaskParameters.getFrameworkParameters();
+
+      // assert
+      expect(parameters.testFiles[0]).to.eq(validFiles[0]);
+    })
+
     it('Should complain if testResultFiles refers to invalid file', () => {
       // arrange
       util.setInput("testResultFormat", "xUnit");     
       util.setInput("testResultFiles", invalidFiles[0]);
       util.loadData();
-      var separator = path.sep;
+      var messageRegex = new RegExp(`Not found testResultFile\\(s\\): .+invalidFile1.xml`);
 
       // act / assert
       require(tp);
-      util.shouldThrow( () => TaskParameters.getFrameworkParameters(), `Not found testResultFile(s): xunit${separator}invalidFile1.xml`);
+      util.shouldThrow( () => TaskParameters.getFrameworkParameters(), messageRegex);
     });
 
     it('Should allow glob paths to be specified', () => {
@@ -177,7 +195,8 @@ describe('TaskParameters', () => {
       // assert
       expect(parameters.testFormat).to.eq("xunit");
       expect(parameters.testFiles.length).to.eq(1);
-
+      // future: make sure that the glob pattern was resolved
+      //expect(parameters.testFiles[0].indexOf("*")).to.eq(-1);
     });
 
     it('Should allow multiple test result files to be specified', () => {

--- a/PublishTestPlanResultsV1/test/testUtil.ts
+++ b/PublishTestPlanResultsV1/test/testUtil.ts
@@ -25,6 +25,7 @@ export function clearData() {
       key.startsWith("VSTS_TASKVARIABLE_") ||
       key.startsWith("BUILD_")
     )
+    // caution: System_* variables should not be deleted
     ).forEach(key => delete process.env[key]);
 }
 

